editor/debugger/debug_adapter/debug_adapter_protocol.cpp

```diff
@@ -671,12 +674,150 @@ int DebugAdapterProtocol::parse_variant(const Variant &p_var) {
+			// Objects have to be requested from the debuggee. This has do be done
+			// in a lazy way, as retrieving object properties takes time.
+			EncodedObjectAsID *encoded_obj = Object::cast_to<EncodedObjectAsID>(p_var);
+
+			// Object may be null; in that case, return early
+			if (!encoded_obj) {
+				return 0;
+			}
+
+			// Object may have been already requested
+			ObjectID object_id = encoded_obj->get_object_id();
+			if (object_list.has(object_id)) {
+				return object_list[object_id];
+			}
+
+			// Queue the object to be requested
+			int id = variable_id++;
+			object_list.insert(object_id, id);
+			return id;
+		}
 		default:
 			// Simple atomic stuff, or too complex to be manipulated
 			return 0;
 	}
 }
 
+void DebugAdapterProtocol::parse_object(SceneDebuggerObject &p_obj) {
+	// If the object is not on the pending list, we weren't expecting it. Ignore it.
+	ObjectID object_id = p_obj.id;
+	if (!object_pending_set.erase(object_id)) {
+		return;
+	}
+
+	// Populate DAP::Variable's with the object's properties. These properties will be divided by categories.
+	Array properties;
+	Array script_members, script_constants, script_node;
+	DAP::Variable node_type;
+	Array node_properties;
+
+	// Lambda to finish the current category
+	auto finishCategory = [&node_properties, &node_type, &properties, this]() {
+		if (!node_properties.is_empty()) {
+			node_type.value = itos(node_properties.size());
+			variable_list.insert(node_type.variablesReference, node_properties.duplicate());
+			properties.push_back(node_type.to_json());
+		}
+	};
+
+	for (SceneDebuggerObject::SceneDebuggerProperty &property : p_obj.properties) {
+		PropertyInfo &info = property.first;
+
+		// Script members ("Members/" prefix)
+		if (info.name.begins_with("Members/")) {
+			info.name = info.name.trim_prefix("Members/");
+			script_members.push_back(parse_object_variable(property));
+		}
+
+		// Script constants ("Constants/" prefix)
+		else if (info.name.begins_with("Constants/")) {
+			info.name = info.name.trim_prefix("Constants/");
+			script_constants.push_back(parse_object_variable(property));
+		}
+
+		// Script node ("Node/" prefix)
+		else if (info.name.begins_with("Node/")) {
+			info.name = info.name.trim_prefix("Node/");
+			script_node.push_back(parse_object_variable(property));
+		}
+
+		// Regular categories (with type Variant::NIL)
+		else if (info.type == Variant::NIL) {
+			finishCategory();
+
+			node_type.name = info.name;
+			node_type.type = "Category";
+			node_type.variablesReference = variable_id++;
+			node_properties.clear();
+		}
+
+		// Regular properties
+		else {
+			node_properties.push_back(parse_object_variable(property));
+		}
+	}
+
+	// Add the last category
+	finishCategory();
+
+	// Add the script categories, in reverse order to be at the front of the array
+	// ( [members; constants; node; category1; category2; ...] )
+	auto addScriptCategories = [&properties, this](const Array &p_properties, const String &p_name) {
```

Comment:
Same goes for [lambdas](https://docs.godotengine.org/en/latest/contributing/development/cpp_usage_guidelines.html#lambdas) unless necessary, this should be a separate method IMO, or use some other means to achieve this

Reply:
My motivation to use lambdas here is essentially:
- Both need to modify local variables (`properties` and `variable_list` in this case), and extracting this to a function and passing in such variables, which have a very specific logic required by the calling function, removes this context and makes it harder to read/modify this function while ensuring it doesn't break something.
- These behaviors are very specific to what this function is doing, and don't seem good candidates to be used in other contexts. Extracting this to a function implies reusable behavior that is generic enough to be reused on other scenarios, which I don't consider to be the case here.

Reply:
If code reduplication is an issue I'd say the better option is just a macro, otherwise this is such a brief snippet and simple that it isn't really a problem IMO to just not use a lambda but add the code directly

Reply:
I think macros would be much worse for this, so sure, since these are small snippets, I'll expand these to not use lambdas :+1:

Reply:
Could be okay here but I think this lies on the "no lambda" side of the boundary (in any case it should be named with `snake_case`)

Reply:
Fixed